Add support for pluck and to_a

kbrock opened the pull request:
You could consider this a bug or an enhancement

Before
======

- `Model.pluck(:a)` was essentially `Model.select("*").to_a.pluck(:a)` -- all fields brought back.

After
=====

- `Model.pluck(:a)` is essentially `Model.select(:a).to_a.pluck(:a)` -- only 1 column brought back.

Extra
=====

- added `Model.to_a`
- added tests ensuring chaining doesn't modify previous step
- added `Model#where_values`
- added yarddoc `@params` for many methods
- fix `where().where()`. It now doesn't modify the intermediate query.

kbrock committed:
use hash for specs

kbrock committed:
add queryable#to_a

This ensures Model.to_a is the same as Model.all.to_a

kbrock committed:
pluck changelog

Fryguy commented:
Are these new or are they bugs because active_support was  removed?

Fryguy commented on lib/query_relation.rb:
```diff
@@ -130,9 +137,22 @@ def offset_value
-  # @param val [Array<Sting,Symbol>,String, Symbol]
-  def select(*args)
-    append_hash_arg :select, *args
+  # @param columns [Array<Sting,Symbol>,String, Symbol] columns to bring back
```
```suggestion
  # @param columns [Array<String, Symbol>, String, Symbol] columns to bring back
```

Fryguy commented on lib/query_relation.rb:
```diff
@@ -130,9 +137,22 @@ def offset_value
+    append_hash_arg :select, *columns
+  end
+
+  # @param columns [Array<Sting,Symbol>,String, Symbol] columns to bring back
```
```suggestion
  # @param columns [Array<String, Symbol>, String, Symbol] columns to bring back
```

kbrock committed:
order use standard argument assignment

Fryguy commented on lib/query_relation.rb:
```diff
@@ -130,9 +137,22 @@ def offset_value
+  end
+
+  # @param columns [Array<Sting,Symbol>,String, Symbol] columns to bring back
+  def pluck(*columns)
```
What happens if someone calls `.pluck` with no arguments?

kbrock replied:
added the use case. looks like pluck is `map(&:values)`

wondering if we want the same implementation

kbrock replied:
active record:

```ruby
VmOrTemplate.limit(1).pluck(:id, :name) 
#   (0.9ms)  SELECT "vms"."id", "vms"."name" FROM "vms" LIMIT $1  [["LIMIT", 1]]
# => [1, "vm"]
VmOrTemplate.limit(1).select(:id, :name, :description).pluck(:id, :name) 
#   (2.0ms)  SELECT "vms"."id", "vms"."name" FROM "vms" LIMIT $1  [["LIMIT", 1]]
# => [1, "vm"]
Vm.limit(1).select(:id, :name, :description).pluck
#   (1.3ms)  SELECT "vms".* FROM "vms" LIMIT $1  [["LIMIT", 1]]
# => [1, "vm", 30 parameters]
```

I feel that last one should have kept the select and had 3 parameters not every parameter from the vm.

UPDATE: added sql in my example to make this more clear

kbrock committed:
ensure chaining methods do not modify original request

kbrock committed:
add where_values;

ensure it is not corrupted by where additions

kbrock committed:
document attributes

kbrock commented:
update:
- fixed yarddocs to add spaces
- add `#where_values`
- added tests for ensuring chaining doesn't modify previous step
- `pluck()` works (not sure if we want this working)
- tweaked `order` implementation (plenty of tests)
- fixed `where` to not modify previous step when 2 hashes are passed in a row.

kbrock commented:
> Are these new or are they bugs because active_support was removed?
> - @Fryguy 

Before removing `ActiveSupport`: `pluck()` had a bad `select()`.
After removing `ActiveSupport`: `pluck()` was dropped and our examples started failing. Adding `ActiveSupport` to the example still has a bad `select()`.

After this PR: `pluck()` is added (yes, implementing more and more of active support interface), and `select()` is correct.

Does look like the edge case `pluck([])` has been fixed, but TBH/ I don't think that is a valid edge case.

Fryguy commented:
Yeah I would expect raising an ArgumentError if pluck is passed an empty list.  Or more accurately I would follow whatever ActiveSupport did

kbrock commented:
I following how `pluck` worked for ActiveRecord models. Again, the way `Model.select(:a).pluck` works was strange, but I followed it

kbrock committed:
pluck

before
======

pluck was handled by enumerable/active support
this did not send select over the wire

after
=====

the pluck attributes are passed via a select

kbrock commented:
I went over the implementation again and realized that `pluck()` just returns an array of empty objects. Removed the special case. (It isn't a valid use case and is unspecified behavior, so no need to write up the code.)

update:
- removed special case for `.pluck()`

kbrock commented:
@Fryguy I'm good with the state of `pluck` -- it is working the same way that active record works.

Any last requests?
